doc_source/falco-fargate-ptrace-security.md

```diff
@@ -0,0 +1,119 @@
+
+Add the following section to the container JSON you wish to secure while creating the task definition. 
+
+```json
+            "linuxParameters": {
+                "capabilities": {
+                    "add": [
+                        "SYS_PTRACE"
+                    ],
+                    "drop": null
+                }
+            },
+```
+
+## Installing Falco
+
+_Note:_ The Falco `pdig` components are targetted to be released in an upcoming Falco release. This documentation will need to be updated after the initial release.
+
+For now you can use the `krisnova/falco-trace:latest` container image which is built on debian and contains Falco and `pdig` pre-installed. 
+The original work to get this working can be found at [github.com/kris-nova/falco-trace](https://github.com/kris-nova/falcotrace).
+
+Follow the official [Falco installation](https://falco.org) guide for installing Falco with `pdig`. 
+
+Validate that Falco can be ran within the container
+
+```bash 
+falco -u 
+
+Sun May 17 13:06:53 2020: Falco initialized with configuration file /etc/falco/falco.yaml
+Sun May 17 13:06:53 2020: Loading rules from file /etc/falco/rules.d/application_rules.yaml:
+Sun May 17 13:06:53 2020: Loading rules from file /etc/falco/rules.d/falco_rules.local.yaml:
+Sun May 17 13:06:53 2020: Loading rules from file /etc/falco/rules.d/falco_rules.yaml:
+^CSun May 17 13:06:59 2020: SIGINT received, exiting...
+Events detected: 0
+Rule counts by severity:
+Triggered rules by rule name:
+Syscall event drop monitoring:
+   - event drop detected: 0 occurrences
+   - num times actions taken: 0
+```
+
+## Running a process with pdig
+
+The above installation should also include a `ptrace(2)` based launcher utility called `pdig`.
+
+The `pdig` utility can either be used to launch a process or attach to a running process. 
+
+Here are some examples of using `pdig` to launch a process.
+
+Running a daemon with `pdig`
+
+```bash
+pdig -a ./app --daemon
+```
+
+Running a program in the background, and attaching to the pid at runtime
+
+```bash
+./app --pidfile=/var/run/app &
+pdig -p $(cat /var/run/app)
+```
+
+Attach to pid 1 with `pdig`. This will trace all pids in the container.
+
+```bash
+pdig -p 1 && ./myapp &
+```
+
+## Logs
+
+In order to consume the Falco logs in ECS the Falco logs must go to STDOUT. 
+
+This can be done a number of ways. One option is to tail the falco log file.
+
+First configure falco to log to a file. One example is [here](https://github.com/kris-nova/falco-trace/blob/master/etc/falco/falco.yaml) or you can edit the file manually.
+
+```bash
+emacs /etc/falco/falco.yaml
+
+# Edit the following
+
+file_output:
+  enabled: true
+  keep_alive: false
+  filename: /var/log/falco.log
+```
+
+After Falco is logging to a file you can run your application.
+
+```bash
+pdig -a ./myapp &
+falco -u --daemon 
+tail -f /var/log/faclo.log &
+```
+
+The logs will now be picked up by ECR and can be used with other AWS services. 
```

Comment:
I don't think the logs will be picked up by ECR... this must be a typo...

Should probably add that you can use FireLens to send the stdout logs somewhere

Reply:
It looks like Falco also supports:

```
stdout_output:
  enabled: true
```

If this is enabled then Falco would just output alerts to stdout directly and FireLens can be used to split the Falco logs out from the application logs and route them to a different alert log group or wherever the customer wants.

@kris-nova Is there a performance advantage or something else I'm missing here that makes it better to use file_output + tail vs just outputting directly to stdout?

Reply:
Hey everyone -

> I don't think the logs will be picked up by ECR... this must be a typo...

Yes. This is a typo :smiley_cat: will fix.

> Should probably add that you can use FireLens to send the stdout logs somewhere

I don't know what FireLens is, so I don't really understand the expectation we want to capture here? Can you suggest something to add to the PR?

---

Regarding `stdout_output.enabled: true` 

That is already [baked into the container image](
https://github.com/kris-nova/falco-trace/blob/master/etc/falco/falco.yaml#L123-L124) I used in the demo and for this writeup. 

I had to tail the logs to effectively break the interactive mode (or at least annoy anyone with a TTY) to get the logs to show up.

I don't know very much about logging in AWS so any help or guidance would be appreciated.

Reply:
Gotcha. Once the updates are made in Falco upstream I'll contribute a sample FireLens task definition config for this. Basically FireLens is a little managed Fluentbit sidecar that runs in your task, and can route the container's logs to different locations based on characteristics of the logs. I think we can make it separate out your application's HTTP access logs and other stdout output from the Falco alert logs themselves, for easier auditing / paging etc. By default ECS will take all the stdout and put it into a single Cloudwatch log group which would make it a bit harder to find the alerts among everything else, but FireLens could split these logs out into two different Cloudwatch log groups for example.

Reply:
Use this as a reference: https://sysdig.com/blog/multi-cluster-security-firelens/